### ISSUE TITLE
How come model works with simpletransformers but not with huggingface?

### ISSUE THREAD

**Issue by drjosephliu** [ThilinaRajapakse/simpletransformers#331]
I'm trying to do multi-class classification over 50 classes on the Reuters 5050 dataset and I'm curious why it works with simpletransformers but not with huggingface. With simpletransformers I get an MCC of 0.59 but with huggingface, I get accuracy of 0.02 (random guessing). 

In fact, when I print out the output, the model outputs the same thing despite different inputs, and it only does this during evaluation.

**With simpletransformers:**

```
train_df = pd.read_pickle(ROOT + 'reuters50_train.pkl')
test_df = pd.read_pickle(ROOT + 'reuters50_test.pkl')

train_df = train_df.drop(columns=['author'])  # column 0 is text and column 1 is author_id
test_df = test_df.drop(columns=['author'])

model = ClassificationModel('bert', 'bert-base-cased', num_labels=50,
                            args={'reprocess_input_data': True, 'overwrite_output_dir': True}) 

model.train_model(train_df)
result, model_outputs, wrong_predictions = model.eval_model(test_df)
print("Result: ", result)
```

**With huggingface:**

```
MAX_LEN = 128
def get_encodings(texts):
    token_ids = []
    attention_masks = []
    for text in texts:
        token_id = tokenizer.encode(text, 
                                    add_special_tokens=True,
                                    max_length=MAX_LEN,
                                    pad_to_max_length=True)
        token_ids.append(token_id)
    return token_ids

def get_attention_masks(padded_encodings):
    attention_masks = []
    for encoding in padded_encodings:
        attention_mask = [int(token_id > 0) for token_id in encoding]
        attention_masks.append(attention_mask)
    return attention_masks


train_df = pd.read_pickle(ROOT + 'reuters50_train.pkl')
test_df = pd.read_pickle(ROOT + 'reuters50_test.pkl')

train_encodings = get_encodings(train_df.text.values)
train_attention_masks = get_attention_masks(train_encodings)

test_encodings = get_encodings(test_df.text.values)
test_attention_masks = get_attention_masks(test_encodings)

X_train = torch.tensor(train_encodings)
y_train = torch.tensor(train_df.author_id.values)
train_masks = torch.tensor(train_attention_masks)

X_test = torch.tensor(test_encodings)
y_test = torch.tensor(test_df.author_id.values)
test_masks = torch.tensor(test_attention_masks)

batch_size = 8

train_data = TensorDataset(X_train, train_masks, y_train)
train_sampler = RandomSampler(train_data)
train_dataloader = DataLoader(train_data, sampler=train_sampler, batch_size=batch_size)

validation_data = TensorDataset(X_test, test_masks, y_test)
validation_sampler = SequentialSampler(validation_data)
validation_dataloader = DataLoader(validation_data, sampler=validation_sampler, batch_size=batch_size)

if torch.cuda.is_available():    
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

config = BertConfig.from_pretrained(
    'bert-base-cased',
    num_labels = 50,
    args={
        'overwrite_output_dir': True,
        'reprocess_input_data': True,
    }
)

model = BertForSequenceClassification(config)

model.to(device)

optimizer = AdamW(model.parameters(),
                  lr = 4e-5, 
                  eps = 1e-8 
                )

from transformers import get_linear_schedule_with_warmup

epochs = 4

total_steps = len(train_dataloader) * epochs

scheduler = get_linear_schedule_with_warmup(optimizer, 
                                            num_warmup_steps = 0, 
                                            num_training_steps = total_steps)


# ========================================
#               Training
# ========================================

loss_values = []

for epoch_i in range(0, epochs):

    t0 = time.time()
    total_loss = 0
    model.train()

    for step, batch in enumerate(train_dataloader):

        b_texts = batch[0].to(device)
        b_attention_masks = batch[1].to(device)
        b_authors = batch[2].to(device)

        model.zero_grad()        

        outputs = model(b_texts, 
                        token_type_ids=None, 
                        attention_mask=b_attention_masks, 
                        labels=b_authors)
        
        loss = outputs[0]

        total_loss += loss.item()

        loss.backward()

        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
        
        optimizer.step()

        scheduler.step()

    avg_train_loss = total_loss / len(train_dataloader)            
    
    loss_values.append(avg_train_loss)

    print("  Average training loss: {0:.2f}".format(avg_train_loss))
    
    # ========================================
    #               Validation
    # ========================================
    # After the completion of each training epoch, measure our performance on
    # our validation set.


    model.eval()

    eval_loss, eval_accuracy = 0, 0
    nb_eval_steps, nb_eval_examples = 0, 0

    for batch in validation_dataloader:
        
        b_texts = batch[0].to(device)
        b_attention_masks = batch[1].to(device)
        b_authors = batch[2].to(device)
        
        with torch.no_grad():        

            outputs = model(b_texts, 
                            token_type_ids=None, 
                            attention_mask=b_attention_masks)
        
        logits = outputs[0]

        logits = logits.detach().cpu().numpy()
        author_ids = b_authors.to('cpu').numpy()
        
        tmp_eval_accuracy = flat_accuracy(logits, author_ids)
        
        eval_accuracy += tmp_eval_accuracy

        nb_eval_steps += 1

    print("  Accuracy: {0:.2f}".format(eval_accuracy/nb_eval_steps))

```

**Comment by ThilinaRajapakse** [comment]
Can you clarify whether the simple transformers model gives you the same output for different inputs, or is it the hugging face model?

**Comment by drjosephliu** [comment]
Ah sorry, I mean to say the huggingface model gives the same output (and thus poor accuracy), whilst your model does as intended. I'm just curious as to why that is because your library interfaces huggingface's and I'm using almost the exact same settings

**Comment by ThilinaRajapakse** [comment]
That's definitely odd. The sequence classification models are identical from what I remember (except possibly how class weights are implemented but you aren't using that anyway). There seems to be a tendency with Transformers models to completely melt down if the learning rate is too high and/or trained for too long. Could this be related to your issue?

**Comment by drjosephliu** [comment]
I've tried a wide range of learning rates.. `[1e-3, 1e-4, 5e-5, 2e-5, 1e-5, 5e-6, 1e-6, 1e-7, 1e-8]`. The results are pretty much uniform across them.

**Comment by ThilinaRajapakse** [comment]
I can't think of any other difference that might be causing this. Are the training loss reduction rates comparable between the two libraries?
